From 6e223cd11fd824362f4caca68e7d456603eb6a01 Mon Sep 17 00:00:00 2001
From: Chris Adams <chris.adams@jollamobile.com>
Date: Thu, 29 Jan 2015 14:01:32 +1000
Subject: [PATCH 6/7] Initialize private member ptr in ctor to avoid crash

This commit initializes the m_secretsDB member of the default secrets
storage class to avoid spurious close() on invalid pointer.

It also ensures that the database name and connection name strings
for the default secrets storage database are detached from their
source strings, in case their source string is a QStringLiteral which
comes from a plugin which gets unloaded prior to the dtor being called.
---
 src/signond/default-secrets-storage.cpp | 10 ++++++----
 src/signond/default-secrets-storage.h   |  1 +
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/signond/default-secrets-storage.cpp b/src/signond/default-secrets-storage.cpp
index b4c0885..00f2b16 100644
--- a/src/signond/default-secrets-storage.cpp
+++ b/src/signond/default-secrets-storage.cpp
@@ -272,7 +272,8 @@ bool SecretsDB::removeData(quint32 id, quint32 method)
 }
 
 DefaultSecretsStorage::DefaultSecretsStorage(QObject *parent):
-    AbstractSecretsStorage(parent)
+    AbstractSecretsStorage(parent),
+    m_secretsDB(0)
 {
 }
 
@@ -288,7 +289,8 @@ bool DefaultSecretsStorage::initialize(const QVariantMap &configuration)
         close();
     }
 
-    QString name = configuration.value(QLatin1String("name")).toString();
+    QString name; // force deep copy / detach
+    name.append(configuration.value(QLatin1String("name")).toString());
 
     m_secretsDB = new SecretsDB(name);
     if (!m_secretsDB->init()) {
@@ -298,6 +300,7 @@ bool DefaultSecretsStorage::initialize(const QVariantMap &configuration)
         return false;
     }
 
+    m_secretsDBConnectionName.append(m_secretsDB->connectionName());
     setIsOpen(true);
     return true;
 }
@@ -305,9 +308,8 @@ bool DefaultSecretsStorage::initialize(const QVariantMap &configuration)
 bool DefaultSecretsStorage::close()
 {
     if (m_secretsDB != 0) {
-        QString connectionName = m_secretsDB->connectionName();
         delete m_secretsDB;
-        QSqlDatabase::removeDatabase(connectionName);
+        QSqlDatabase::removeDatabase(m_secretsDBConnectionName);
         m_secretsDB = 0;
     }
     return AbstractSecretsStorage::close();
diff --git a/src/signond/default-secrets-storage.h b/src/signond/default-secrets-storage.h
index 0ef9372..d8c254a 100644
--- a/src/signond/default-secrets-storage.h
+++ b/src/signond/default-secrets-storage.h
@@ -95,6 +95,7 @@ public:
 
 private:
     SecretsDB *m_secretsDB;
+    QString m_secretsDBConnectionName;
 };
 
 } //namespace
-- 
2.20.1

